packages/apollo-server-core/src/ApolloServer.ts

```diff
@@ -512,27 +522,44 @@ export class ApolloServerBase {
         `called start() with surprising state ${initialState.phase}`,
       );
     }
+    const maybeGateway =
+      initialState.phase === 'initialized with gateway'
+        ? initialState.gateway
+        : undefined;
     const barrier = resolvable();
     this.state = { phase: 'starting', barrier };
     let loadedSchema = false;
     try {
       const schemaDerivedData =
         initialState.phase === 'initialized with schema'
           ? initialState.schemaDerivedData
-          : this.generateSchemaDerivedData(
-              await this.startGatewayAndLoadSchema(initialState.gateway),
-            );
+          : this.generateSchemaDerivedData({
+              schema: await this.startGatewayAndLoadSchema(
+                initialState.gateway,
+              ),
+            });
       loadedSchema = true;
       this.state = {
         phase: 'invoking serverWillStart',
         barrier,
-        schemaDerivedData,
+        // We can't just use the variable schemaDerivedData here, as any
```

Comment:
I mean, this complexity is exactly why I wanted to differentiate between initial load and later updates, and why my instinct was to move away from "use the same callback for initial load and later update". I feel like the old logic of "it's the job of gateway.load to ensure that at the moment it resolves its promise, that's the current schema, and as long as you don't yield between `await load()` and updating the state and starting to listen for changes, it works out.

I don't want to jerk you back and forth here... but I thought the point of doing onSchemaLoadOrUpdate was to make this stuff less complex, not more complex.  This is really hard to follow right now.

Reply:
>I mean, this complexity is exactly why I wanted to differentiate between initial load and later updates, and why my instinct was to move away from "use the same callback for initial load and later update". I feel like the old logic of "it's the job of gateway.load to ensure that at the moment it resolves its promise, that's the current schema, and as long as you don't yield between `await load()` and updating the state and starting to listen for changes, it works out.

I'm not sure separating initial load and later updates would help for this specific problem. Specifically, the following can happen:
- While executing `_start()`, `await gateway.load()` runs, which executes `gateway.load()`.
- At some point during the execution of `gateway.load()`, it returns a promise that resolves to the initial schema, and execution suspends while waiting on that promise. Let's suppose the promise resolves at time A.
- At some point later (time B), `_start()` resumes execution (now that the promise has resolved, execution is unblocked).
- Between A and B, it's possible for other code to execute, including the code that updates the schema.

Basically it's hard to keep the initial schema non-stale when it's dynamic.

>I don't want to jerk you back and forth here... but I thought the point of doing onSchemaLoadOrUpdate was to make this stuff less complex, not more complex. This is really hard to follow right now.

FWIW, I think once AS3 lands and we can require gateways to define `onSchemaLoadOrUpdate`, this changes to a:
```typescript
this.state = {
  // ...
  schemaDerivedData: this.state.schemaDerivedData,
}
```
like the other changes to `this.state`, and `this.state.schemaDerivedData` effectively becomes the source of truth.

Reply:
Maybe we could do a (noninvasive) "polyfill"? Like we can define a wrapper "gateway" which, if gateway.onSchemaLoadOrUpdate doesn't exist, we wrap around the gateway. It passes through methods other than onSchemaLoadOrUpdate and load, and handles onSchemaLoadOrUpdate by delegating to onSchemaChange and also keeping it around to call after load?  So we could put all the back compat logic in this one wrapper class, and keep the main startup logic clean?

Reply:
@glasser I ended up going with something a bit more general than a polyfill here, it's at `packages/apollo-server-core/src/utils/schemaManager.ts` (it wraps gateway and provides `onSchemaLoadOrUpdate`, but it also wraps schemas in the non-gateway case and is responsible for updating schema-derived data).